File: backend/src/main/java/com/interview/controller/UserController.java
package com.interview.controller;

import com.interview.dto.ApiResponse;
import com.interview.dto.ChangePasswordRequest;
import com.interview.dto.UpdateProfileRequest;
import com.interview.dto.UserProfileDTO;
import com.interview.service.UserService;
import com.interview.util.SecurityUtils;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

/**
 * REST controller for user management operations
 */
@Slf4j
@RestController
@RequestMapping("/api/v1/users")
@RequiredArgsConstructor
@Tag(name = "User Management", description = "User Management API")
public class UserController {

    private final UserService userService;
    private final SecurityUtils securityUtils;

    @GetMapping("/profile")
    @Operation(summary = "Get user profile")
    public ResponseEntity<ApiResponse<UserProfileDTO>> getProfile() {
        log.info("GET /api/v1/users/profile");
        String userId = securityUtils.getCurrentUserId();
        UserProfileDTO profile = userService.getUserProfile(userId);
        return ResponseEntity.ok(ApiResponse.success(profile));
    }

    @PutMapping("/profile")
    @Operation(summary = "Update user profile")
    public ResponseEntity<ApiResponse<UserProfileDTO>> updateProfile(
            @Valid @RequestBody UpdateProfileRequest request) {
        log.info("PUT /api/v1/users/profile");
        String userId = securityUtils.getCurrentUserId();
        UserProfileDTO profile = userService.updateProfile(userId, request);
        return ResponseEntity.ok(ApiResponse.success("Profile updated successfully", profile));
    }

    @PostMapping("/change-password")
    @Operation(summary = "Change user password")
    public ResponseEntity<ApiResponse<String>> changePassword(
            @Valid @RequestBody ChangePasswordRequest request) {
        log.info("POST /api/v1/users/change-password");
        String userId = securityUtils.getCurrentUserId();
        userService.changePassword(userId, request);
        return ResponseEntity.ok(ApiResponse.success("Password changed successfully", null));
    }

    @PostMapping("/profile-image")
    @Operation(summary = "Upload profile image")
    public ResponseEntity<ApiResponse<UserProfileDTO>> uploadProfileImage(
            @RequestParam("file") MultipartFile file) {
        log.info("POST /api/v1/users/profile-image");
        String userId = securityUtils.getCurrentUserId();
        UserProfileDTO profile = userService.uploadProfileImage(userId, file);
        return ResponseEntity.ok(ApiResponse.success("Profile image uploaded successfully", profile));
    }

    @PostMapping("/skills")
    @Operation(summary = "Add skill")
    public ResponseEntity<ApiResponse<String>> addSkill(@RequestParam String skill) {
        log.info("POST /api/v1/users/skills - Add skill: {}", skill);
        String userId = securityUtils.getCurrentUserId();
        userService.addSkill(userId, skill);
        return ResponseEntity.ok(ApiResponse.success("Skill added successfully", null));
    }

    @DeleteMapping("/skills")
    @Operation(summary = "Remove skill")
    public ResponseEntity<ApiResponse<String>> removeSkill(@RequestParam String skill) {
        log.info("DELETE /api/v1/users/skills - Remove skill: {}", skill);
        String userId = securityUtils.getCurrentUserId();
        userService.removeSkill(userId, skill);
        return ResponseEntity.ok(ApiResponse.success("Skill removed successfully", null));
    }
}

File: backend/src/controllers/user.controller.js
const userService = require('../services/user.service');
const ApiResponse = require('../utils/api-response');
const { getCurrentUserId } = require('../utils/security.utils');

const getProfile = async (req, res, next) => {
  try {
    const userId = getCurrentUserId(req);
    const profile = await userService.getUserProfile(userId);
    return res.status(200).json(ApiResponse.success(profile));
  } catch (error) {
    next(error);
  }
};

const updateProfile = async (req, res, next) => {
  try {
    const userId = getCurrentUserId(req);
    const profile = await userService.updateProfile(userId, req.body);
    return res.status(200).json(ApiResponse.success('Profile updated successfully', profile));
  } catch (error) {
    next(error);
  }
};

const changePassword = async (req, res, next) => {
  try {
    const userId = getCurrentUserId(req);
    await userService.changePassword(userId, req.body);
    return res.status(200).json(ApiResponse.success('Password changed successfully', null));
  } catch (error) {
    next(error);
  }
};

const uploadProfileImage = async (req, res, next) => {
  try {
    if (!req.file) {
      return res.status(400).json(ApiResponse.error('File parameter is required'));
    }
    const userId = getCurrentUserId(req);
    const profile = await userService.uploadProfileImage(userId, req.file);
    return res.status(200).json(ApiResponse.success('Profile image uploaded successfully', profile));
  } catch (error) {
    next(error);
  }
};

const addSkill = async (req, res, next) => {
  try {
    const userId = getCurrentUserId(req);
    const skill = req.query.skill || req.body.skill;
    if (!skill) {
      return res.status(400).json(ApiResponse.error('Skill parameter is required'));
    }
    await userService.addSkill(userId, skill);
    return res.status(200).json(ApiResponse.success('Skill added successfully', null));
  } catch (error) {
    next(error);
  }
};

const removeSkill = async (req, res, next) => {
  try {
    const userId = getCurrentUserId(req);
    const skill = req.query.skill || req.body.skill;
    if (!skill) {
      return res.status(400).json(ApiResponse.error('Skill parameter is required'));
    }
    await userService.removeSkill(userId, skill);
    return res.status(200).json(ApiResponse.success('Skill removed successfully', null));
  } catch (error) {
    next(error);
  }
};

module.exports = {
  getProfile,
  updateProfile,
  changePassword,
  uploadProfileImage,
  addSkill,
  removeSkill
};
